src/workerd/api/sockets.c++

```diff
@@ -160,7 +160,9 @@ jsg::Promise<void> Socket::close(jsg::Lock& js) {
 jsg::Ref<Socket> Socket::startTls(jsg::Lock& js, jsg::Optional<TlsOptions> tlsOptions) {
   JSG_REQUIRE(!isSecureSocket, TypeError, "Cannot startTls on a TLS socket.");
   // TODO: Track closed state of socket properly and assert that it hasn't been closed here.
-  JSG_REQUIRE(domain != nullptr, TypeError, "startTLS can only be called once.");
+  JSG_REQUIRE(domain != nullptr, TypeError, "startTls can only be called once.");
+  JSG_REQUIRE(!readable->getPipeToCalled(), TypeError,
+      "Cannot call startTls after `pipeTo` was used.");
```

Comment:
Not giong to block but I'm really not a fan of this. While it is most likely that startTls will be called very close to the start of the stream after only a few bytes are read, there are conceivable cases where I might pipe the stream to something else to do the actual handling of the data. For instance, what if I want to introduce a logging TransformStream transparently into the mix... e.g.

```
const { readable, writable } = socket;
const r = readable.pipeThrough(new LoggingTransform());
const reader = r.getReader();
await reader.read(); // do just a single read...
socket.startTls(); // fails!
```

I still might only pull a few of the initial bytes off the reader to process then try performing the startTls operation, which will now fail.

Reply:
We either need to do this, or we need to make the application specify at startup whether or not it plans to do startTls() later. If we don't do either of these things, then regular old stream-proxying use cases will end up hard to optimize.

Reply:
Based on our discussion I thought we were in agreement to do this. Is there a better way?

But actually as things stand I think your example will work, only `ReadableStream::pipeTo` sets this flag. I was actually wondering whether we should disable on `pipeThrough` as well?

Reply:
If we do this `pipeThrough` and `pipeTo` should both have the same effect. As I said, I'm not going to block on this, I'm just not convinced it's the right thing to do. Personally I'm fine with an option on connect to say that `startTls` is not expected.

Reply:
Yeah, I understand the hesitation.

What it comes down to for me is that `pipeTo`/`pipeThrough` disabling `startTls` is a concrete functionality change which people will notice, whereas an option on `connect` which doesn't change functionality but improves performance is likely to be missed (majority of folks will get the slow path because they won't realise that they need to disable startTls).

I think it's preferable to have an option in the future which makes the code slower but enables `pipeTo`+`startTls`. This will force people to explicitly choose which path they desire.

Reply:
Perhaps consider adding that option now?

Reply:
My personal preference is to require the app to specify its intent to use starttls at connect time, with the default being that starttls is _off_, i.e. an opt-in approach. I don't like an opt-out approach because most developers probably won't notice it and so won't get the benefit. Also remember that it's not just for the benefit of the app, but also the benefit of the platform.

I don't think people like the opt-in approach, so my second choice is the one in this PR, which I think will do the right thing in the vast majority of cases.

Note that if we discover this PR causes problems in practice, we can always undo it later. Removing this restriction would be a backwards-compatible change, but adding it is backwards-incompatible, so if we're unsure if we need it we should add it now.

Reply:
I would be +1 on the explicit opt-in option set at connect time.

Reply:
I think I'm the one getting in the way here. Sounds like everyone above thinks a `startTls` option is ideal. Reading the thread I think I see why, and how an opt-in option could work.

```typescript
import { connect } from "cloudflare:sockets"

// Works
const socket = connect(address, {startTls: true});
socket.startTls();

// Fails with error — default value of startTls option is *false*
const socket = connect(address);
socket.startTls();
// Error copy: "Cannot call startTls() on a socket that was not provided the startTls option when it was created. To fix this error, set the startTls option to true when calling connect(). Example: connect(address, {startTls: true});
```

Still feels a bit odd, but as long as we have very clear error text, think we can live with it.

Reply:
> I don't think people like the opt-in approach

I think the same. By taking this approach we are leaking our implementation details directly into the API. I also think other runtimes might not be happy to adopt this, so standardisation will be tougher.

Reply:
@dom96:
> I also think other runtimes might not be happy to adopt this...

I might be misunderstanding but if this means that you think other runtimes would be unhappy with having a `startTls` enable/disable *option* in the connect call, I definitely disagree. From Node.js' point of view, an explicit option to disable startTls from the start would be significantly better than relying on whether or not a stream happens to be piped to/from at any given point. In Node.js we've been explicitly trying to move away from adding non-standard behaviors to standardized API implementations and that would include attaching additional internal properties to ReadableStream/WritableStream to track something like this.

@irvinebroque:
>> const socket = connect(address);
>> socket.startTls();
>> // Error copy: "Cannot call startTls() on a socket that was not provided the allowStartTls option when it was created. To fix > this error, set the allowStartTls option to true when calling connect(). Example: connect(address, {allowStartTls: true});
> Still feels a bit odd, but as long as we have very clear error text, think we can live with it.

Truthfully, with my Node.js developer hat on, I do not consider this to be odd at all. It is perfectly reasonable and something that we can easily explain, especially given that we're going to need a similar option for starting the connection with TLS from the *beginning*. There's *very little* difference between `connect(address, { useSecureTransport: true })` and `connect(address, { allowStartTls: true })` ... 

In fact, just bikeshedding a bit, these options should be named similarly to be consistent:

* `useSecureTransport` -- use TLS from the beginning...
* `allowSecureTransport` -- enable the startTls API...

Or maybe even combine them into an enum option rather than separate booleans:

```
connect(address, { secureTransport: 'use' });  // use tls from the beginning...
connect(address, { secureTransport: 'allow' });  // allow startTls....
```

Reply:
Okay, I'm won over James' suggestion. Shall I create a quick PR instead to add the three-way `secureTransport` enum? (`off`, `use`, `allow`)

Reply:
+1, I'm sold. Nice one @jasnell

Reply:
SGTM as long as default is "off" (so that we can optimize appropriately).